### ISSUE TITLE
perf(compactor): Improve task measurement on the compactor side with parallelism

### ISSUE THREAD

**Issue by Li0k** [risingwavelabs/risingwave#13811]
### Background

1. Compactor uses a `pull-based` architecture and uses `cpu_core` to estimate the number of tasks a node can load. 
2. With the `fast compact` turned on in the compactor,  the CPU usage of each task decreases, and more is blocked on the network IO instead of the CPU.
3. Only a single quota is calculated for tasks executed in parallel.

### Improve

1. Due to the `fast compact`, the CPU consumption of each task is low, we should increase the ratio to improve the service capacity of the compactor.
2. Consider parallelism in quota calculation.
3. When the quota is not enough to serve the parallelism of a new task, refuse to execute the new task.

**Comment by Li0k** [comment]
@hzxa21 @Little-Wallace How do you feel about this issue ?

**Comment by hzxa21** [comment]
> 1. Due to the fast compact, the CPU consumption of each task is low, we should increase the ratio to improve the service capacity of the compactor.
> 2. Consider parallelism in quota calculation.
> 3. When the quota is not enough to serve the parallelism of a new task, refuse to execute the new task.

IMO, these are valid improvements. I think we should focus on how to implement them. Questions:
- For 1, how to decide the ratio?
- For 2, how to calculate the parallelism of a task? IIRC, the parallelism calculation somteimes relies on block meta to be fetched from s3.
-  For 3, doesn't the refusion logic already exist?

**Comment by Li0k** [comment]
1. Determining the ratio directly through CPU or other metrics seems to be inaccurate because the consumption of CPU by compact tasks is unpredictable (IO), any suggestions? 
2. Based on 1, calculating the parallelism of a task on meta might be a better approach. Currently, when the compactor pulls a task, the parallelism of the task is regarded as 1. Under the new strategy, it may have a negative impact and I will reduce the number of tasks pulled each time.
3. The resource control strategy relies on memory denial. As mentioned in 2, the CPU estimation is not accurate. We expect to limit the CPU usage through pull task count.

**Comment by Li0k** [comment]
> * Determining the ratio directly through CPU or other metrics seems to be inaccurate because the consumption of CPU by compact tasks is unpredictable (IO), any suggestions?
> * Based on 1, calculating the parallelism of a task on meta might be a better approach. Currently, when the compactor pulls a task, the parallelism of the task is regarded as 1. Under the new strategy, it may have a negative impact and I will reduce the number of tasks pulled each time.
> * The resource control strategy relies on memory denial. As mentioned in 2, the CPU estimation is not accurate. We expect to limit the CPU usage through pull task count.

Determining the ratio directly through CPU or other metrics seems to be inaccurate because the consumption of CPU by compact tasks is unpredictable (IO), any suggestions?
Based on 1, calculating the parallelism of a task on meta might be a better approach. Currently, when the compactor pulls a task, the parallelism of the task is regarded as 1. Under the new strategy, it may have a negative impact and I will reduce the number of tasks pulled each time.
The resource control strategy relies on memory denial. As mentioned in 2, the CPU estimation is not accurate. We expect to limit the CPU usage through pull task count.

**Comment by hzxa21** [comment]
> 1. Determining the ratio directly through CPU or other metrics seems to be inaccurate because the consumption of CPU by compact tasks is unpredictable (IO), any suggestions?
> 2. Based on 1, calculating the parallelism of a task on meta might be a better approach. Currently, when the compactor pulls a task, the parallelism of the task is regarded as 1. Under the new strategy, it may have a negative impact and I will reduce the number of tasks pulled each time.
> 3. The resource control strategy relies on memory denial. As mentioned in 2, the CPU estimation is not accurate. We expect to limit the CPU usage through pull task count.


Does meta have sufficient information to determine the task parallelism? I am thinking whether it is better to simply let compactor pull more tasks (equivalent to increase the ratio) and reject tasks if its CPU usage is consistently higher than a threshold.

**Comment by Li0k** [comment]
> > 1. Determining the ratio directly through CPU or other metrics seems to be inaccurate because the consumption of CPU by compact tasks is unpredictable (IO), any suggestions?
> > 2. Based on 1, calculating the parallelism of a task on meta might be a better approach. Currently, when the compactor pulls a task, the parallelism of the task is regarded as 1. Under the new strategy, it may have a negative impact and I will reduce the number of tasks pulled each time.
> > 3. The resource control strategy relies on memory denial. As mentioned in 2, the CPU estimation is not accurate. We expect to limit the CPU usage through pull task count.
> 
> Does meta have sufficient information to determine the task parallelism? I am thinking whether it is better to simply let compactor pull more tasks (equivalent to increase the ratio) and reject tasks if its CPU usage is consistently higher than a threshold.

1. If we keep the logic consistent with the compactor, the meta needs to access the object store to get enough information which is what we want to avoid.
2. In line with the point of view of this issue, we simplified the problem to pulling more tasks and preventing CPU overload with a rejection policy. However, the statistics of CPU load may be delayed. If we use the CPU to reject tasks, will it cause a period of delay?  I don't have a good idea, should we let the compactor run out of resources or stay below a threshold?
